pkg/pubsub/sqs/subscriber.go

```diff
@@ -0,0 +1,99 @@
+package sqs
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
+
+	"github.com/scribd/go-sdk/pkg/pubsub"
+	"github.com/scribd/go-sdk/pkg/pubsub/pool"
+)
+
+type (
+	Subscriber struct {
+		client      *sqs.Client
+		queueURL    string
+		handler     MsgHandler
+		maxMessages int
+		pool        *pool.Pool
+		waitTime    time.Duration
+	}
+
+	SubscriberConfig struct {
+		SQSClient  *sqs.Client
+		MsgHandler MsgHandler
+		SQSConfig  pubsub.SQS
+	}
+
+	MsgHandler func(msg types.Message)
+)
+
+const (
+	defaultNumWorkers = 1
+	maxWaitTime       = 20 * time.Second
+)
+
+func NewSubscriber(c SubscriberConfig) *Subscriber {
+	workers := c.SQSConfig.Subscriber.Workers
+	if workers == 0 {
+		workers = defaultNumWorkers
+	}
+
+	waitTime := c.SQSConfig.Subscriber.WaitTime
+	if waitTime > maxWaitTime {
+		waitTime = maxWaitTime
+	}
+
+	return &Subscriber{
+		client:      c.SQSClient,
+		handler:     c.MsgHandler,
+		maxMessages: c.SQSConfig.Subscriber.MaxMessages,
+		queueURL:    c.SQSConfig.Subscriber.QueueURL,
+		pool:        pool.New(workers),
+		waitTime:    waitTime,
+	}
+}
+
+func (s *Subscriber) Subscribe(ctx context.Context) chan error {
+	ch := make(chan error)
+
+	req := &sqs.ReceiveMessageInput{
+		QueueUrl:              aws.String(s.queueURL),
+		MaxNumberOfMessages:   int32(s.maxMessages),
+		MessageAttributeNames: []string{"All"},
+		AttributeNames:        []types.QueueAttributeName{"All"},
+	}
+	if s.waitTime > 0 {
+		req.WaitTimeSeconds = int32(s.waitTime.Seconds())
+	}
+
+	go func() {
+		defer close(ch)
+
+		for {
+			response, err := s.client.ReceiveMessage(ctx, req)
+
+			if err != nil {
+				ch <- err
+
+				return
+			}
+
+			for _, message := range response.Messages {
+				s.pool.Schedule(func() {
+					s.handler(message)
+				})
+			}
+		}
+
+	}()
+
+	return ch
+}
+
+func (s *Subscriber) Unsubscribe() error {
+	return nil
```

Comment:
Why don't we use the same approach as in Kafka? Putting the messages in a goroutine, and then close the channel here?

Reply:
@laynax It is a good idea to have a signal channel for that, I have added it ✅